crates/containerd-shim-wasm/src/sandbox/exec.rs

```diff
@@ -0,0 +1,252 @@
+}
+
+impl From<Status> for ExitStatus {
+    fn from(status: Status) -> Self {
+        match status {
+            Status::Exited(status) => status,
+            Status::Running => panic!("cannot convert running process to exit status"),
+        }
+    }
+}
+
+fn wait(fd: RawFD, no_hang: bool) -> Result<Status, Error> {
+    let no_hang = if no_hang {
+        WaitPidFlag::WNOHANG
+    } else {
+        WaitPidFlag::empty()
+    };
+
+    let info = waitid(WaitID::PIDFd(fd), no_hang | WaitPidFlag::WEXITED)
+        .map_err(|e| std::io::Error::from(e))?;
+
+    match info {
+        WaitStatus::Exited(pid, status) => Ok(Status::Exited(ExitStatus {
+            pid: pid.as_raw() as u32,
+            status: status as u32,
+        })),
+        WaitStatus::Signaled(pid, sig, dumped) => {
+            info!("child {} killed by signal {}, dumped: {}", pid, sig, dumped);
+            Ok(Status::Exited(ExitStatus {
+                pid: pid.as_raw() as u32,
+                status: 128 + sig as u32,
+            }))
+        }
+        WaitStatus::StillAlive => Ok(Status::Running),
+        _ => Err(Error::Others(format!("unexpected wait status: {:?}", info))),
+    }
+}
+
+impl PidFD {
+    // wait for the process referred to by the pidfd to exit
+    //
+    // If you want more control over waiting you can use `as_raw_fd()` and call `waitid` yourself.
+    pub fn wait(&self) -> Result<ExitStatus, Error> {
+        let ws = wait(self.fd, false)?;
+        Ok(ws.into())
+    }
+
+    pub fn wait_no_hang(&self) -> Result<Status, Error> {
+        wait(self.fd, true)
+    }
+
+    // Send the specified signal to the process referred to by the pidfd.
+    pub fn kill(&self, sig: i32) -> Result<(), Error> {
+        let ret = unsafe { pidfd_send_signal(self.fd, sig, ptr::null_mut(), 0) };
+        if ret == -1 {
+            return Err(std::io::Error::last_os_error().into());
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+pub struct ExitStatus {
+    pub pid: u32,
+    pub status: u32,
+}
+
+pub enum Context {
+    // Parent stores the pid of the child process and the pidfd that can be used to, for instance, wait on the child.
+    Parent(u32, PidFD),
+    Child,
+}
+
+pub fn has_cap_sys_admin() -> bool {
+    let caps = caps::read(None, CapSet::Effective).unwrap();
+    caps.contains(&Capability::CAP_SYS_ADMIN)
+}
+
+// This is is a wrapper around clone3 which allows us to pass a pidfd
+// This works otherwise just like normal forking semantics:
+// If this is the parent, the result will be Ok(Context::Parent(pid, pidfd)), where the pid is the pid of the new process.
+// If this is the child, the result will be Ok(Context::Child).
+//
+// Code that runs in the child must not do things like access locks or other shared state.
+//
+// Optionally you can pass in a reference to a file descriptor which will be populated with the pidfd (see pidfd_open(2)).
+pub unsafe fn fork(cgroup: Option<&Box<dyn Cgroup>>) -> Result<Context, Error> {
```

Comment:
```suggestion
pub fn fork(cgroup: Option<&dyn Cgroup>) -> Result<Context, Error> {
```

If we are passing a reference, we don't need the `Box` around this -- will suggest modifications to the two function calls of this too

Reply:
Oh great, I thought this wasn't quite right.
I see now I can pass the boxed reference with `.as_ref()`.